### DISCUSSION TITLE
Get a list of available state files

### DISCUSSION THREAD

**Discussion by designermonkey** [gruntwork-io/knowledge-base#683] [How do I]
I'm following best practices for splitting state out into multiple files which results in the following in my s3 bucket:

```
- /my-project/nyc1/dns.terraform.tfstate
- /my-project/nyc1/uat/dns.terraform.tfstate
- /my-project/nyc1/uat/mysql.terraform.tfstate
- /my-project/nyc1/uat/vpc.terraform.tfstate
- /my-project/nyc1/uat/instance/app.terraform.tfstate
- /my-project/nyc1/uat/instance/dns.terraform.tfstate
- /my-project/nyc1/uat/instance/firewalls.terraform.tfstate
- /my-project/nyc1/uat/instance/ingress.terraform.tfstate
- /my-project/nyc1/uat/instance/route.terraform.tfstate
- /my-project/nyc1/uat/instance/ssh.terraform.tfstate
- /my-project/nyc1/uat/instance/ssl.terraform.tfstate
```

I need to use these files to access output values in my ansible project. I can easily access the files themselves, but the issue I have is knowing what file is where. I have a question:

Has anyone got good advice on how to get a manifest of all the state files, maybe as a terraform module I can run, which stores a state file at `/my-project/manifest.terraform.tfstate` for example?

Maybe I'm barking up the wrong tree and need to rethink how I get the state in ansible...

---

<ins datetime="2023-03-16T17:04:47Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109992">Tracked in ticket #109992</a></p>
</ins>

**Comment by pras111gw** [comment]
Hi @designermonkey can you explain the use case? If you are looking to read outputs of your resources (a vpc id, for example), an output parameter is a better choice.

**Comment by ryehowell** [comment]
Hi @designermonkey,

This is a great question and is a good use case as to why having a good naming/pathing convention really helps. There is a great [blog post by Yevgeniy Brikman](https://blog.gruntwork.io/how-to-manage-terraform-state-28f5697e68fa) that covers "Isolation via file layout" (feel free to skip to that section of the post) which is what I always try to follow in my Terraform projects. It looks like your state file layout follows a similar pattern. A single manifest files typically wouldn't be needed if the state files follow a well known pattern and naming convention. This way I can templatize the path to the state file that will work across environment lifecycles and use cases. It can be as simple or complex as needed depending on the resource footprint.

For example, if I have an Ansible Role that is going to configure my web-server with a DB connection string, then I can templatize the path to the state file to get the DB info (or anything else that I need from Terraform outputs/state): `{state-bucket-name}/{environment}/{service}/terraform.tfstate`, where the template values would be well-known such as `prod` or `stage` for the environment, and `mysql` or `database` for the service. This is a very simple example and it can be much more complex as needed.

This does depend on "how" the state files are being split and how the overall project structure is.
